sandbox/reflection/struct.h

```diff
@@ -0,0 +1,192 @@
+  }                                                                                                           \
+  template <class F>                                                                                          \
+  static void CURRENT_REFLECTION(F&& f, bricks::reflection::Index<bricks::reflection::FieldName, idx>) {      \
+    f(#name);                                                                                                 \
+  }                                                                                                           \
+  template <class F>                                                                                          \
+  static void CURRENT_REFLECTION(F&& f,                                                                       \
+                                 bricks::reflection::Index<bricks::reflection::FieldTypeAndName, idx>) {      \
+    f(bricks::reflection::TypeWrapper<type>(), #name);                                                        \
+  }                                                                                                           \
+  template <class F>                                                                                          \
+  void CURRENT_REFLECTION(F&& f, bricks::reflection::Index<bricks::reflection::FieldValue, idx>) {            \
+    f(name);                                                                                                  \
+  }                                                                                                           \
+  template <class F>                                                                                          \
+  void CURRENT_REFLECTION(                                                                                    \
+      F&& f, bricks::reflection::Index<bricks::reflection::FieldNameAndImmutableValueReference, idx>) const { \
+    f(#name, name);                                                                                           \
+  }                                                                                                           \
+  template <class F>                                                                                          \
+  void CURRENT_REFLECTION(                                                                                    \
+      F&& f, bricks::reflection::Index<bricks::reflection::FieldNameAndMutableValueReference, idx>) {         \
+    f(#name, name);                                                                                           \
+  }
+
+#define CURRENT_STRUCT_IS_VALID(s) \
+  bricks::reflection::CurrentStructFieldsConsistency<s, bricks::reflection::FieldCounter<s>::value - 1>::Check()
+
+namespace bricks {
+namespace reflection {
+
+template <typename T>
+struct FieldCounter {
+  static_assert(std::is_base_of<CurrentBaseType, T>::value,
+                "Template argument must be derived from `CurrentBaseType`.");
```

Comment:
This error message explains the problem but not why it happened or how to fix it.